Restore _Float128 support by default

smowton opened the pull request:
_Float128 was exposed by GCC as of version 4.3 (x86-64) to 4.5 (i386 and I think others).
cbmc still makes no effort to imitate a particular GCC version, so I simply enable recognising
the token by default, which used to happen universally; goto-cc on the other hand uses the
new `gcc_versiont` to choose an appropriate behaviour.

This fixes a regression on tests using math.h due to some versions of glibc defining _Float128
conditional on the compiler version, which cbmc and goto-cc now both allow to appear like the
gcc found on the path.

The excerpt from libc causing the problem:
```
#if (defined __x86_64__                                                 \
     ? __GNUC_PREREQ (4, 3)                                             \
     : (defined __GNU__ ? __GNUC_PREREQ (4, 5) : __GNUC_PREREQ (4, 4))) \
     &&  !defined(__CUDACC__) && !defined(__ICC)
# define __HAVE_FLOAT128 1
```

Unsure what's going on with `__GNU__`, which I can't find any definition for, but 4.3+ for x86-64 and 4.5+ for other platforms seems reasonable. This arises now because the use-default-compiler-defines PR was merged, meaning cbmc no longer always claims gcc 4.2.1.

smowton committed:
Restore _Float128 support by default

_Float128 was exposed by GCC as of version 4.3 (x86-64) to 4.5 (i386 and I think others).
cbmc still makes no effort to imitate a particular GCC version, so I simply enable recognising
the token by default, which used to happen universally; goto-cc on the other hand uses the
new `gcc_versiont` to choose an appropriate behaviour.

This fixes a regression on tests using math.h due to some versions of glibc defining _Float128
conditional on the compiler version, which cbmc and goto-cc now both allow to appear like the
gcc found on the path.

tautschnig commented:
We really need some regression test here. This seems to be the third or fourth closely-related PR around this topic, and we need to ensure that we are monotone in the sense of having fewer bugs, not more.

tautschnig commented on src/goto-cc/gcc_mode.cpp:
```diff
@@ -530,6 +530,13 @@ int gcc_modet::doit()
+    config.ansi_c.arch == "x86_64" ? 3 : 5;
+
+  config.ansi_c.Float128_type =
+    gcc_version.flavor==gcc_versiont::flavort::GCC &&
```
~I believe Clang also supports it.~

tautschnig replied:
clang-format suggests space around `==`

tautschnig commented on src/util/config.cpp:
```diff
@@ -912,6 +912,9 @@ bool configt::set(const cmdlinet &cmdline)
     ansi_c.preprocessor=ansi_ct::preprocessort::GCC;
   }
 
+  if(ansi_c.preprocessor == ansi_ct::preprocessort::GCC)
```
Clang?

smowton replied:
Agreed, I'll try to cook one up tomorrow morning...

smowton commented:
Re: clang, `_Float128` support appears to have been briefly added (https://reviews.llvm.org/D40673) then reverted (https://github.com/llvm-mirror/clang/commit/2b1d36bee69eaf8897af89d53630804f9f66badd) in January (heading towards Clang-6.0 release), so I don't think we need to handle this as a built-in Clang type yet. Contrast `__float128`, which it recognises as of at least clang-4.0.

smowton committed:
Improve gcc-4 and -7 tests

These now check all ts18661 types, and the wrapper scripts avoid argument re-parsing

smowton committed:
Add test for gcc-5

This version should expose _Float128, but not other ts18661 types

smowton committed:
Extend cbmc ts18661 test

This now expects to find _Float128 when gcc-4.3+ (x86-64) for -4.5+ (other targets) is in use.

smowton commented:
@tautschnig existing tests extended to examine the intermediate (gcc >= 4.3, < 7.0) position where _Float128 is available but _Float32 et al are not.

`cbmc` is currently acting oddly -- it will mimic gcc-5 behaviour (define _Float128, don't define _Float32) regardless of the preprocessor it uses. That will likely break gcc <4.3 with a modern libc that fills in _Float128 (whether the type is used or not), and gcc >= 7.0 trying to use _Float32, but that's the same brokenness it exhibited before any of the TS18661 stuff started. The solution is for cbmc to act more like goto-cc (learn our version from the real gcc, then act like it).

tautschnig commented on regression/ansi-c/gcc_version1/fake-gcc-5:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+gcc -Wno-macro-redefined -U __clang_major__ -D __GNUC__=5 -D __GNUC_MINOR__=0 -D __GNUC_PATCHLEVEL__=0 "$@"
+
```
Nit-pick: no need for blank line.

tautschnig commented on regression/ansi-c/gcc_version1/gcc-5.c:
```diff
@@ -0,0 +1,11 @@
+// But this type should:
+_Float128 f128;
+
+
```
Nit-pick: no need for blank lines.

tautschnig commented on regression/cbmc/ts18661_typedefs/main.c:
```diff
@@ -1,18 +1,30 @@
+
 #if __GNUC__ >= 7
 #define HAS_FLOATN
+#elif __GNUC_PREREQ(4, FLOAT128_MINOR_VERSION)
```
The failing test suggests that `__GNUC_PREREQ` isn't universally available.

tautschnig commented on src/ansi-c/ansi_c_language.cpp:
```diff
@@ -76,6 +76,7 @@ bool ansi_c_languaget::parse(
   ansi_c_parser.set_message_handler(get_message_handler());
   ansi_c_parser.for_has_scope=config.ansi_c.for_has_scope;
   ansi_c_parser.ts_18661_3_Floatn_types=config.ansi_c.ts_18661_3_Floatn_types;
+  ansi_c_parser.Float128_type=config.ansi_c.Float128_type;
```
Nit-pick: space around `=`

tautschnig commented on src/goto-cc/gcc_mode.cpp:
```diff
@@ -530,6 +530,13 @@ int gcc_modet::doit()
     config.ansi_c.ts_18661_3_Floatn_types=true;
 
+  int gcc_float128_minor_version =
+    config.ansi_c.arch == "x86_64" ? 3 : 5;
```
clang-format suggests that all of this fits in one line.

tautschnig reviewed (commented):
Playing the clang-format echo machine...

smowton commented:
@tautschnig comments addressed

kroening commented:
@smowton The plan is to make cbmc behave exactly like goto-cc.

kroening commented:
To add, it should behave like goto-cl on Windows, of course.
